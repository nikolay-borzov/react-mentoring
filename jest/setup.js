import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
