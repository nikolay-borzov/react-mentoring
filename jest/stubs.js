export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre'],
  vote_average: 7.5
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
