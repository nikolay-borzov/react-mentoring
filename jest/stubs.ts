import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
