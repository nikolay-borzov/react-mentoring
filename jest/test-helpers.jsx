import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
