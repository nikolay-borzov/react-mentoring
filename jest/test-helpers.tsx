import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
