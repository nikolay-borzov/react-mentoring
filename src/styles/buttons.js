// @flow

import React from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

import { HoverEffectMixin } from './mixins'

export const Button = styled.input`
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  transition: background-color var(--transition-duration),
              box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  user-select: none;

  ${HoverEffectMixin}

  ${props =>
    props.small
      ? `padding: var(--padding-button-small);
         font-size: 1rem;`
      : ''}

  ${props => (props.primary ? `background-color: var(--color-primary);` : '')}
`

// https://github.com/styled-components/styled-components/issues/1198
export const NavigateButton = Button.withComponent(
  ({ primary, small, ...rest }) => <Link {...rest} />
).extend`
  ${props =>
    props.primary ? `color: var(--color-text-light) !important;` : ''}
`
