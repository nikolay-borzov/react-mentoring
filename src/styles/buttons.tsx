import React from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

import { HoverEffectMixin } from './mixins'

interface ButtonProps {
  primary?: boolean
  small?: boolean
}

export const Button = styled.input<ButtonProps>`
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  transition: background-color var(--transition-duration),
              box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  user-select: none;

  ${HoverEffectMixin}

  ${props =>
    props.small
      ? `padding: var(--padding-button-small);
         font-size: 1rem;`
      : ''}

  ${props => (props.primary ? `background-color: var(--color-primary);` : '')}
`

export const NavigateButton = styled(props => <Button as={Link} {...props} />)`
  ${props =>
    props.primary ? `color: var(--color-text-light) !important;` : ''}
`
