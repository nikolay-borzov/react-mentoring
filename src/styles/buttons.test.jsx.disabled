import React from 'react'
import 'jest-styled-components'

import { itRendersCorrectlyShallow } from '../../jest/test-helpers'

import { Button, NavigateButton } from './buttons'

describe('ButtonStyled', () => {
  itRendersCorrectlyShallow(() => {
    return <Button />
  }, 'default')

  itRendersCorrectlyShallow(() => {
    return <Button primary />
  }, 'primary')

  itRendersCorrectlyShallow(() => {
    return <Button small />
  }, 'small')
})

describe('NavigateButtonStyled', () => {
  itRendersCorrectlyShallow(() => {
    return <NavigateButton />
  }, 'default')

  itRendersCorrectlyShallow(() => {
    return <NavigateButton primary />
  }, 'primary')

  itRendersCorrectlyShallow(() => {
    return <NavigateButton small />
  }, 'small')
})
