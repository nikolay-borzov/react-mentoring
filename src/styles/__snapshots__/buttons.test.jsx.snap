// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonStyled default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<input
  className="c0"
/>
`;

exports[`ButtonStyled primary 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  background-color: var(--color-primary);
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<input
  className="c0"
/>
`;

exports[`ButtonStyled small 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  padding: var(--padding-button-small);
  font-size: 1rem;
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<input
  className="c0"
/>
`;

exports[`NavigateButtonStyled default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<Component
  className="c0"
/>
`;

exports[`NavigateButtonStyled primary 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  background-color: var(--color-primary);
  color: var(--color-text-light) !important;
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<Component
  className="c0"
  primary={true}
/>
`;

exports[`NavigateButtonStyled small 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: var(--padding-button);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  transition: background-color var(--transition-duration),box-shadow var(--transition-duration);
  padding: var(--padding-button-small);
  font-size: 1rem;
}

.c0:hover {
  box-shadow: 0 0 20px var(--color-primary),2px 2px 2px rgba(0,0,0,0.5);
}

<Component
  className="c0"
  small={true}
/>
`;
